### PR TITLE
Build the restrictor from diagonal entries

### LINKED ISSUE #97: System Matrix required when building the restrictor
When computing the restrictor [here](https://github.com/ORNL-CEES/mfmg/blob/master/include/mfmg/dealii/amge_host.hpp#L78), we require the `system_sparse_matrix`. This is not necessary, what we really need it the diagonal. This is pretty standard to compute it in a matrix-free manner (see [here](
https://dealii.org/developer/doxygen/deal.II/classMatrixFreeOperators_1_1Base.html))

### PR DESCRIPTION
**Pull request by dalg24**
Fix #97

### CODE REVIEW

File: tests/test_restriction_matrix.cc [hunk 1]
```diff
@@ -128,10 +128,12 @@ BOOST_AUTO_TEST_CASE(restriction_matrix)
     system_sparse_matrix.set(index, index, 1.0);
   system_sparse_matrix.compress(dealii::VectorOperation::insert);
 
+  auto locally_relevant_global_diag = evaluator.get_diagonal();
+
   dealii::TrilinosWrappers::SparseMatrix restriction_sparse_matrix;
   amge.compute_restriction_sparse_matrix(
       eigenvectors, diag_elements, dof_indices_maps, n_local_eigenvectors,
-      system_sparse_matrix, restriction_sparse_matrix);
+      locally_relevant_global_diag, restriction_sparse_matrix);
```
Comment (dalg24) [inline tests/test_restriction_matrix.cc]: @Rombur This yield an error:
```bash
qdi@dgx-dl01:/tmp/mfmg/build/tests$ ./test_restriction_matrix --run_test=weight_sum
Running 1 test case...

*** No errors detected
qdi@dgx-dl01:/tmp/mfmg/build/tests$ ./test_restriction_matrix --run_test=restriction_matrix
Running 1 test case...

--------------------------------------------------------
An error occurred in line <160> of file </scratch/source/dealii/source/base/partitioner.cc> in function
    void dealii::Utilities::MPI::Partitioner::set_ghost_indices(const dealii::IndexSet&, const dealii::IndexSet&)
The violated condition was:
    ghost_indices_in.n_elements() == 0 || ghost_indices_in.size() == locally_owned_range_data.size()
Additional information:
    Dimension 25 not equal to 0.

Stacktrace:
-----------
#0  /opt/dealii/lib/libdeal_II.g.so.9.1.0-pre: dealii::Utilities::MPI::Partitioner::set_ghost_indices(dealii::IndexSet const&, dealii::IndexSet const&)
#1  /opt/dealii/lib/libdeal_II.g.so.9.1.0-pre: dealii::Utilities::MPI::Partitioner::Partitioner(dealii::IndexSet const&, dealii::IndexSet const&, ompi_communicator_t*)
#2  /opt/dealii/lib/libdeal_II.g.so.9.1.0-pre: dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host>::reinit(dealii::IndexSet const&, dealii::IndexSet const&, ompi_communicator_t*)
#3  /opt/dealii/lib/libdeal_II.g.so.9.1.0-pre: dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host>::Vector(dealii::IndexSet const&, dealii::IndexSet const&, ompi_communicator_t*)
#4  ./test_restriction_matrix: ) [0x58ed8e]
#5  ./test_restriction_matrix: ) [0x4d1eae]
#6  ./test_restriction_matrix: ) [0x4d1402]
#7  ./test_restriction_matrix: ) [0x4f9edb]
#8  /opt/boost/lib/libboost_unit_test_framework.so.1.67.0: boost::detail::function::function_obj_invoker0<boost::detail::forward, int>::invoke(boost::detail::function::function_buffer&)
#9  /opt/boost/lib/libboost_unit_test_framework.so.1.67.0: boost::execution_monitor::catch_signals(boost::function<int ()> const&)
#10  /opt/boost/lib/libboost_unit_test_framework.so.1.67.0: boost::execution_monitor::execute(boost::function<int ()> const&)
#11  /opt/boost/lib/libboost_unit_test_framework.so.1.67.0: boost::execution_monitor::vexecute(boost::function<void ()> const&)
#12  /opt/boost/lib/libboost_unit_test_framework.so.1.67.0: boost::unit_test::unit_test_monitor_t::execute_and_translate(boost::function<void ()> const&, unsigned int)
#13  /opt/boost/lib/libboost_unit_test_framework.so.1.67.0: boost::unit_test::framework::state::execute_test_tree(unsigned long, unsigned int, boost::unit_test::framework::state::random_generator_helper const*)
#14  /opt/boost/lib/libboost_unit_test_framework.so.1.67.0: boost::unit_test::framework::state::execute_test_tree(unsigned long, unsigned int, boost::unit_test::framework::state::random_generator_helper const*)
#15  /opt/boost/lib/libboost_unit_test_framework.so.1.67.0: boost::unit_test::framework::run(unsigned long, bool)
#16  /opt/boost/lib/libboost_unit_test_framework.so.1.67.0: boost::unit_test::unit_test_main(bool (*)(), int, char**)
#17  ./test_restriction_matrix: ) [0x4d1103]
#18  /lib/x86_64-linux-gnu/libc.so.6: __libc_start_main
#19  ./test_restriction_matrix: ) [0x4d0b49]
--------------------------------------------------------

[dgx-dl01:07668] *** Process received signal ***
[dgx-dl01:07668] Signal: Aborted (6)
[dgx-dl01:07668] Signal code:  (-6)
[dgx-dl01:07668] [ 0] /lib/x86_64-linux-gnu/libc.so.6(+0x354b0)[0x7f6e86b234b0]
[dgx-dl01:07668] [ 1] /lib/x86_64-linux-gnu/libc.so.6(gsignal+0x38)[0x7f6e86b23428]
[dgx-dl01:07668] [ 2] /lib/x86_64-linux-gnu/libc.so.6(abort+0x16a)[0x7f6e86b2502a]
[dgx-dl01:07668] [ 3] /opt/dealii/lib/libdeal_II.g.so.9.1.0-pre(_ZN6dealii18deal_II_exceptions9internals22do_issue_error_nothrowERKNS_13ExceptionBaseE+0x0)[0x7f6e96e44852]
[dgx-dl01:07668] [ 4] ./test_restriction_matrix(_ZN6dealii18deal_II_exceptions9internals20issue_error_noreturnINS_18StandardExceptions20ExcDimensionMismatchEEEvNS1_17ExceptionHandlingEPKciS7_S7_S7_T_+0x82)[0x4e0659]
[dgx-dl01:07668] [ 5] /opt/dealii/lib/libdeal_II.g.so.9.1.0-pre(_ZN6dealii9Utilities3MPI11Partitioner17set_ghost_indicesERKNS_8IndexSetES5_+0xcb)[0x7f6e96f47101]
[dgx-dl01:07668] [ 6] /opt/dealii/lib/libdeal_II.g.so.9.1.0-pre(_ZN6dealii9Utilities3MPI11PartitionerC2ERKNS_8IndexSetES5_P19ompi_communicator_t+0x26f)[0x7f6e96f4925f]
[dgx-dl01:07668] [ 7] /opt/dealii/lib/libdeal_II.g.so.9.1.0-pre(_ZN6dealii13LinearAlgebra11distributed6VectorIdNS_11MemorySpace4HostEE6reinitERKNS_8IndexSetES8_P19ompi_communicator_t+0x46)[0x7f6e96925b2c]
[dgx-dl01:07668] [ 8] /opt/dealii/lib/libdeal_II.g.so.9.1.0-pre(_ZN6dealii13LinearAlgebra11distributed6VectorIdNS_11MemorySpace4HostEEC2ERKNS_8IndexSetES8_P19ompi_communicator_t+0x12b)[0x7f6e96925cd3]
[dgx-dl01:07668] [ 9] ./test_restriction_matrix[0x58ed8e]
[dgx-dl01:07668] [10] ./test_restriction_matrix[0x4d1eae]
[dgx-dl01:07668] [11] ./test_restriction_matrix[0x4d1402]
[dgx-dl01:07668] [12] ./test_restriction_matrix[0x4f9edb]
[dgx-dl01:07668] [13] /opt/boost/lib/libboost_unit_test_framework.so.1.67.0(_ZN5boost6detail8function21function_obj_invoker0INS0_7forwardEiE6invokeERNS1_15function_bufferE+0x2c)[0x7f6e9b4cb8bc]
[dgx-dl01:07668] [14] /opt/boost/lib/libboost_unit_test_framework.so.1.67.0(_ZN5boost17execution_monitor13catch_signalsERKNS_8functionIFivEEE+0x14d)[0x7f6e9b4cac7d]
[dgx-dl01:07668] [15] /opt/boost/lib/libboost_unit_test_framework.so.1.67.0(_ZN5boost17execution_monitor7executeERKNS_8functionIFivEEE+0x58)[0x7f6e9b4cad68]
[dgx-dl01:07668] [16] /opt/boost/lib/libboost_unit_test_framework.so.1.67.0(_ZN5boost17execution_monitor8vexecuteERKNS_8functionIFvvEEE+0x56)[0x7f6e9b4cb4e6]
[dgx-dl01:07668] [17] /opt/boost/lib/libboost_unit_test_framework.so.1.67.0(_ZN5boost9unit_test19unit_test_monitor_t21execute_and_translateERKNS_8functionIFvvEEEj+0x441)[0x7f6e9b4fa381]
[dgx-dl01:07668] [18] /opt/boost/lib/libboost_unit_test_framework.so.1.67.0(_ZN5boost9unit_test9framework5state17execute_test_treeEmjPKNS2_23random_generator_helperE+0x636)[0x7f6e9b4d6226]
[dgx-dl01:07668] [19] /opt/boost/lib/libboost_unit_test_framework.so.1.67.0(_ZN5boost9unit_test9framework5state17execute_test_treeEmjPKNS2_23random_generator_helperE+0x8fb)[0x7f6e9b4d64eb]
[dgx-dl01:07668] [20] /opt/boost/lib/libboost_unit_test_framework.so.1.67.0(_ZN5boost9unit_test9framework3runEmb+0x480)[0x7f6e9b4cdfc0]
[dgx-dl01:07668] [21] /opt/boost/lib/libboost_unit_test_framework.so.1.67.0(_ZN5boost9unit_test14unit_test_mainEPFbvEiPPc+0x3e7)[0x7f6e9b4f82c7]
[dgx-dl01:07668] [22] ./test_restriction_matrix[0x4d1103]
[dgx-dl01:07668] [23] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf0)[0x7f6e86b0e830]
[dgx-dl01:07668] [24] ./test_restriction_matrix[0x4d0b49]
[dgx-dl01:07668] *** End of error message ***
Aborted (core dumped)
```

Comment (dalg24) [reply tests/test_restriction_matrix.cc]: This was due to the use of `DummyMeshEvaluator` and was resolved by extracting the diagonal entries directly from `system_sparse_matrix`.

File: include/mfmg/common/amge.hpp [hunk 2]
```diff
@@ -67,7 +67,6 @@ class AMGe
    */
   void output(std::string const &filename) const;
 
-protected:
```
Comment (dalg24) [inline include/mfmg/common/amge.hpp]: Has to be public if we want to call from `test_restriction_matrix.cc`

File: include/mfmg/dealii/dealii_mesh_evaluator.hpp [hunk 3]
```diff
@@ -58,9 +60,9 @@ class DealIIMeshEvaluator : public MeshEvaluator
     ASSERT_THROW_NOT_IMPLEMENTED();
   }
 
-  dealii::DoFHandler<dim> &get_dof_handler();
+  dealii::DoFHandler<dim> &get_dof_handler() const;
```
Comment (Rombur) [inline include/mfmg/dealii/dealii_mesh_evaluator.hpp]: Isn't that kind of strange to have a `const` function that return a `non-const` reference to protected data ?

Comment (dalg24) [reply include/mfmg/dealii/dealii_mesh_evaluator.hpp]: Good point.  Let me revisit that one.

Comment (dalg24) [reply include/mfmg/dealii/dealii_mesh_evaluator.hpp]: Considering that `DealIIMeshEvaluator::evaluate_{agglomerate, global}()` require non-`const` reference to the dof handler and the constraints, I think the right thing to do is revert my marking `get_dof_handler()` and `get_constraints()` as `const` member function and have `get_diagonal()` just not be `const`-qualified.

Comment (Rombur) [reply include/mfmg/dealii/dealii_mesh_evaluator.hpp]: That's fine.

**Comment by Rombur** [comment]
Instead of reverting the commits you drop them? That would simplify the history.

**Comment by dalg24** [comment]
> Instead of reverting the commits you drop them? That would simplify the history.

Done

Did you want to see more history cleanup?

**Comment by Rombur** [comment]
If it is possible, can you remove the commits where you implemented everything in `HierarchyHelpers`. Otherwise looks good.

**Pull request merged**